### ISSUE TITLE
enable daemons/pmdas logging to stderr

### ISSUE THREAD

**Issue by fche** [performancecopilot/pcp#365]
In a containerized or systemd environments, the canonical way for processes to produce logs is to print them to stderr.  That way, errors are not lost to log files made transient by container restarts, or by the various little "foo.log" -> "foo.log.old" file-renaming mechanisms.

pmcd, its pmdas, and other daemonic tools should have readily accessible 'log straight to stderr' modes.

**Comment by fche** [comment]
(This is a more urgent part of issue #177.)

**Comment by goodwinos** [comment]
We could enhance libpcp_pmda:: pmdaOpenLog(pmdaInterface * dispatch) with some new syntax in the dispatch->version.any.ext->e_logfile field to specify host/port/protocol of a remote or centralized logging host. By this I mean pmcd, pmlogger and certain PMDAs that normally call pmdaOpenLog to freopen stderr to a local file in $PCP_LOG_DIR/pmcd/foo.log. The new functionality could instead send these messages to a remote host, perhaps via an rsyslog plugin or something.

**Comment by fche** [comment]
In a container environment, there may not be a remote host to log to.  The simplest thing - to send messages to stderr, would be enough. (For pipe pmda processes, the pmcd connection eats up stdin/stdout.)

**Comment by goodwinos** [comment]
Well, that would mean checking if dispatch->version.any.ext->e_logfile is "stderr" and NOT doing the freopen dance. But where is stderr going to go in that case? The PMDA is launched by pmcd and stderr will be pmcd.log if it's a DSO.

**Comment by fche** [comment]
> But where is stderr going to go in that case?

Along with pmcd's own logging output, it should go to pmcd's original stderr.

**Comment by fche** [comment]
The simplest way may be to support an environment variable that, if set, makes pmdaOpenLog and/or __pmOpenLog just return a dup(stderr) filehandle instead of opening a file.

**Comment by goodwinos** [comment]
With pmcd and potentially multiple PMDAs all logging to stderr, we'd need line buffering (stderr does that anyway) but also each message prefixed with the PMDA domain name, not just the number .. otherwise it'd just be a jumble of messages in the one log. I don't think we have a pmda domain number to name translation table, so that might be needed (could read stdpmid)

**Comment by natoscott** [comment]
@goodwinos the PMDAs know their own names, so should be no need for domain numbers and/or lookup tables here.  FWIW, pminfo can also potentially see the PMDA error messages (in local context mode) so an audit of those messages would be helpful in general.

**Comment by goodwinos** [comment]
Proposed patch - have tested this with pmcd -f -l stderr but no formal QA yet.

```
commit 86885ce27aeb9919c6f0f6cdb3b39cda876bb330
Author: Mark Goodwin <mgoodwin@redhat.com>
Date:   Fri Apr 27 14:25:51 2018 +1000

    libpcp: allow pmOpenLog() to log to "stderr"
    
    GH #365 - enable daemons/pmdas logging to stderr
    
    Treat "stderr" specially as the log file argument to pmOpenLog().
    modified:   src/libpcp/src/util.c

diff --git a/src/libpcp/src/util.c b/src/libpcp/src/util.c
index 87cc7cd97..ec1ae0b8c 100644
--- a/src/libpcp/src/util.c
+++ b/src/libpcp/src/util.c
@@ -361,6 +361,16 @@ FILE *
 pmOpenLog(const char *progname, const char *logname, FILE *oldstream,
            int *status)
 {
+    if (oldstream == stderr && logname && strcmp(logname, "stderr") == 0) {
+       /*
+        * Special case to just write to existing stderr stream,
+        * no need to reopen or dup anything.
+        */
+       logheader(progname, oldstream, "started");
+       *status = 0;
+       return oldstream;
+    }
+
     oldstream = logreopen(progname, logname, oldstream, status);
     logheader(progname, oldstream, "started");
 
@@ -388,6 +398,11 @@ __pmRotateLog(const char *progname, const char *logname, FILE *oldstream,
     int                i;
     FILE       *newstream = oldstream;
 
+    if (oldstream == stderr && logname && strcmp(logname, "stderr") == 0) {
+       /* special case, as for pmOpenLog(), see above */
+       return oldstream;
+    }
+
     PM_LOCK(util_lock);
     for (i = 0; i < nfilelog; i++) {
        if (oldstream == filelog[i])
```

**Comment by natoscott** [comment]
@goodwinos I guess I'd maybe be inclined to use the string "-" instead of "stderr" ... otherwise seems fine to me. pmOpenLog(3) will need some explanation around this too.

**Comment by goodwinos** [comment]
thanks @natoscott I'm happy to use "-", or "stderr", or even "" or NULL. Whatever everyone agrees on, or all of the above. The use case for this is in a container environment where pmcd is started in the background to the main pid-1 application, and needs to log to stderr. See https://github.com/redhat-developer/osd-monitor-poc/issues/13

BTW, in the proposed patch, the second hunk for __pmRotateLog() should set *status = 0 the same as the first hunk does.

**Comment by natoscott** [comment]
@goodwinos something that will not conflict with any existing file name (or potential argument) - so I agree, NULL or an empty string are other good options here.

**Comment by fche** [comment]
Does the draft patch above automatically cover the case of PMDAs (pmdaOpenLog calling down to pmOpenLog incidentally)?  We'd like to merge in their outputs into the overall pmcd stderr stream too.

**Comment by goodwinos** [comment]
pmdaOpenLog() calls pmOpenLog(), so yes - provided the pmdaInterface *dispatch argument specifies dispatch->version.any.ext->e_logfile = "-" (or whatver special name we decide on .. I like "-").
In practice, this means the Install script for the PMDA would have to specify args="-l -" to override the default log filename, or as in the osd-monitor-poc containers, just specify it directly in the pmcd.conf constructed by the Dockerfile. This assumes the PMDA supports the -l option (pmdaprometheus does not - I'll add it).

I have only tested DSOs with this so far, as per the background shell script used in various Dockerfile in osd-monitor-poc

**Comment by fche** [comment]
Does your draft of the code implement the pmda-name-prefixing part we were discussing back up at https://github.com/performancecopilot/pcp/issues/365#issuecomment-351229843?

**Comment by goodwinos** [comment]
Yes, for well behaved daemon PMDAs, you get the name and PID of the pmda process prefixing all messages. And the messages are line buffered. For example, with the following parameters in pmcd.conf for the trivial and prometheus agents -

```
prometheus      144     pipe    binary notready         python3 /var/lib/pcp/pmdas/prometheus/pmdaprometheus.python -l-
trivial 250     pipe    binary          /var/lib/pcp/pmdas/trivial/pmdatrivial -d 250 -l- -Dlibpmda
```

i.e. -l- (minus l minus says use stderr), and I've also added -Dlibpmda for trivial, we get log messages like the following in pmcd.log -

```
Log for pmcd on goblin started Wed May  2 09:47:09 2018

Log for pmdaprometheus on goblin started Wed May  2 09:47:10 2018

[Wed May  2 09:47:10] pmdaprometheus(5091) Info: Initializing ... currently in notready state.
[Wed May  2 09:47:10] pmdaprometheus(5091) Info: Config change detected, traversed 2 config entries in 0.0001s, rescanning ...
[Wed May  2 09:47:10] pmdaprometheus(5091) Info: Found source source1 cluster 1
[Wed May  2 09:47:10] pmdaprometheus(5091) Info: Ready to process requests
Log for pmdatrivial on goblin started Wed May  2 09:47:10 2018

[Wed May  2 09:47:10] pmdatrivial(5099) Debug: pmdaInit: PMDA pmdatrivial: help file /var/lib/pcp/pmdas/trivial/help opened
[Wed May  2 09:47:10] pmdatrivial(5099) Info: name        = pmdatrivial
[Wed May  2 09:47:10] pmdatrivial(5099) Info: domain      = 250
[Wed May  2 09:47:10] pmdatrivial(5099) Info: ext flags  = 4
[Wed May  2 09:47:10] pmdatrivial(5099) Info: num metrics = 1
[Wed May  2 09:47:10] pmdatrivial(5099) Info: num indom   = 0
[Wed May  2 09:47:10] pmdatrivial(5099) Info: metric map  = direct
[Wed May  2 09:47:10] pmdatrivial(5099) Debug: pmdaConnect: PMDA pmdatrivial: opened pipe to pmcd, infd = 0, outfd = 1
...
[Wed May  2 09:47:10] pmcd(5082) Info: PMNS file "DEFAULT" is unchanged
```
Of course if pmcd itself was run in the foreground e.g. pmcd -f -l-   then all pmcd messages and all messages from PMDAs using stderr, will go to pmcd's stderr stream, which can be redirected to wherever you want in a container environment. PMDAs that have **not** specified -l- will write their log messages to the default location, which is $PCP_LOG_DIR/pmcd/PMDANAME.log

**Comment by fche** [comment]
> then all pmcd messages and all messages from PMDAs using stderr, will go to pmcd's stderr stream

perfect

**Comment by goodwinos** [comment]
Fixed a long time ago with 23b79e0 and 7fdf7b3. Supporting` -l-` for log messages to stdout has been especially useful.
